solr/solrj/src/java/org/apache/solr/client/solrj/request/JavaBinUpdateRequestCodec.java

```diff
@@ -277,67 +208,52 @@ public List<Object> readIterator(DataInputInputStream fis) throws IOException {
-
-          if (o == END_OBJ) {
-            break;
-          }
+    private void readDocs(DataInputInputStream fis) throws IOException {
+      if (resultNamedList == null) resultNamedList = new NamedList<>();
 
-          SolrInputDocument sdoc = null;
-          if (o instanceof List) {
-            @SuppressWarnings("unchecked")
-            List<NamedList<?>> list = (List<NamedList<?>>) o;
-            sdoc = listToSolrInputDocument(list);
-          } else if (o instanceof NamedList) {
-            UpdateRequest req = new UpdateRequest();
-            req.setParams(new ModifiableSolrParams(((NamedList) o).toSolrParams()));
-            handler.update(null, req, null, null);
-          } else if (o instanceof Map.Entry) {
-            @SuppressWarnings("unchecked")
-            Map.Entry<SolrInputDocument, Map<?, ?>> entry =
-                (Map.Entry<SolrInputDocument, Map<?, ?>>) o;
-            sdoc = entry.getKey();
-            Map<?, ?> p = entry.getValue();
-            if (p != null) {
-              commitWithin = (Integer) p.get(UpdateRequest.COMMIT_WITHIN);
-              overwrite = (Boolean) p.get(UpdateRequest.OVERWRITE);
-            }
-          } else if (o instanceof SolrInputDocument) {
-            sdoc = (SolrInputDocument) o;
-          } else if (o instanceof Map) {
-            sdoc = convertMapToSolrInputDoc((Map) o);
-          }
+      UpdateRequest updateRequest = new UpdateRequest();
+      NamedList<?> params = (NamedList<?>) resultNamedList.get("params"); // always precedes docs
+      if (params != null) {
+        updateRequest.setParams(ModifiableSolrParams.of(params.toSolrParams()));
+      }
 
-          // peek at the next object to see if we're at the end
-          o = readVal(fis);
-          if (o == END_OBJ) {
-            // indicate that we've hit the last doc in the batch, used to enable optimizations when
-            // doing replication
-            updateRequest.lastDocInBatch();
+      Object o = readVal(fis);
+      while (o != END_OBJ) {
+        Integer commitWithin = null;
+        Boolean overwrite = null;
+
+        SolrInputDocument sdoc;
+        if (o instanceof Map.Entry) { // doc + options.  UpdateRequest "docsMap"
+          @SuppressWarnings("unchecked")
+          Map.Entry<SolrInputDocument, Map<?, ?>> entry =
+              (Map.Entry<SolrInputDocument, Map<?, ?>>) o;
+          sdoc = entry.getKey();
+          Map<?, ?> p = entry.getValue();
+          if (p != null) {
+            commitWithin = (Integer) p.get(UpdateRequest.COMMIT_WITHIN);
+            overwrite = (Boolean) p.get(UpdateRequest.OVERWRITE);
           }
+        } else if (o instanceof SolrInputDocument d) { // doc.  UpdateRequest "docs""
+          sdoc = d;
+        } else if (o instanceof Map<?, ?> m) { // doc.  To imitate JSON style.  SOLR-13731
+          sdoc = convertMapToSolrInputDoc(m);
+        } else {
+          throw new IllegalStateException("Unexpected data type: " + o.getClass());
```

Comment:
Not being silent about the unexpected!  I'll change this to `SolrException(ErrorCode.BAD_REQUEST`